Public/Src/Engine/Processes/SandboxedProcessReports.cs

```diff
@@ -647,6 +648,44 @@ private bool FileAccessReportLineReceived<T>(ref T data, FileAccessReportProvide
                 }
             }
 
+            if (operation == ReportedFileOperation.FirstAllowWriteCheckInProcess)
+            {
+                // This operation represents that a given path was checked for write access for the first time
+                // within the scope of a process. The status of the operation represents whether that access should
+                // have been allowed/denied, based on the existence of the file.
+                // However, we need to determine whether to deny the access based on the first time the path was
+                // checked for writes across the whole process tree. This means checking the first time this operation 
+                // is reported for a given path, and ignore subsequent reports.
+                // Races are ignored: a race means two child processes are racing to create or delete the same file
+                // - something that is not a good build behavior anyway - and the outcome will be that we will 
+                // non -deterministically deny the access
+                if (path != null && !m_overrideAllowedWritePaths.ContainsKey(path))
+                {
+                    // We should override write allowed accesses for this path if the status of the special operation was 'denied'
+                    m_overrideAllowedWritePaths[path] = (status == FileAccessStatus.Denied);
+                }
+
+                return true;
+            }
+
+            FileAccessStatusMethod method;
+
+            // If we are processing an allowed write, but this should be overridden based on file existence, 
+            // we change the status here
+            if (path != null &&
+                (requestedAccess & RequestedAccess.Write) != 0 &&
+                status == FileAccessStatus.Allowed &&
+                m_overrideAllowedWritePaths.TryGetValue(path, out bool shouldOverrideAllowedAccess) &&

```

Comment:
>m_overrideAllowedWritePaths [](start = 16, length = 27)

If `TryGetValue` doesn't check for size first, then I suggest that you check `m_overrideAllowedWritePaths.Count != 0`.

Reply:
It's a regular dictionary, I'd expect that to cost the same...

---
In reply to: [299188507](https://github.com/microsoft/BuildXL/pull/536#discussion_r299188507) [](ancestors = 299188507)